### ISSUE TITLE
Clippy lint "comparison to empty" fails to apply fix when reference is used

### ISSUE THREAD

**Issue by mdgaziur** [rust-lang/rust#104888]
I tried this code:

```rust
fn main() {
    let s = "Hello World!";
    let s1 = &s;
    println!("{}", *s1 != "");
}
```

I expected to see this happen: Clippy applies fix successfully

Instead, this happened: 

Clippy fails to automatically apply fix

<details><summary>Output</summary>
<p>

```
> cargo clippy --fix
    Checking clippy_bug v0.1.0 (/home/mdgaziur001/clippy_bug)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_bug`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0614]: type `bool` cannot be dereferenced
 --> src/main.rs:4:20
  |
4 |     println!("{}", *s1.is_empty());
  |                    ^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0614`.
Original diagnostics will follow.

warning: comparison to empty slice
 --> src/main.rs:4:20
  |
4 |     println!("{}", *s1 == "");
  |                    ^^^^^^^^^ help: using `is_empty` is clearer and more explicit: `*s1.is_empty()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#comparison_to_empty
  = note: `#[warn(clippy::comparison_to_empty)]` on by default

warning: `clippy_bug` (bin "clippy_bug") generated 1 warning (run `cargo fix --bin "clippy_bug"` to apply 1 suggestion)
warning: `clippy_bug` (bin "clippy_bug" test) generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 1.20s
```
</p>
</details>

### Meta

`rustc --version --verbose`:

Stable:

```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

Beta:

```
rustc 1.66.0-beta.2 (0040709e3 2022-11-20)
binary: rustc
commit-hash: 0040709e39bfa159d099ee14628ad4ac72f91406
commit-date: 2022-11-20
host: x86_64-unknown-linux-gnu
release: 1.66.0-beta.2
LLVM version: 15.0.2
```

Nightly:

```
rustc 1.66.0-beta.2 (0040709e3 2022-11-20)
binary: rustc
commit-hash: 0040709e39bfa159d099ee14628ad4ac72f91406
commit-date: 2022-11-20
host: x86_64-unknown-linux-gnu
release: 1.66.0-beta.2
LLVM version: 15.0.2
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by mdgaziur** [comment]
I can think of two ways clippy can fix this:
```rust
(*s1).is_empty()
```

or

```rust
s1.is_empty()
```

**Comment by mdgaziur** [comment]
I can think of two ways clippy can fix this:
```rust
(*s1).is_empty()
```

or

```rust
s1.is_empty()
```

**Comment by mdgaziur** [comment]
I have made a fix but I'm not sure where to open pull request

**Comment by fmease** [comment]
> I have made a fix but I'm not sure where to open pull request

As far as I know, over at [rust-lang/rust-clippy](github.com/rust-lang/rust-clippy). Further, this issue should probably be moved to that repo as well (by maintainers).
